src/model.ts

```diff
@@ -90,23 +123,52 @@ export abstract class Model<T extends Model<T>> {
       await data.remove(this.__snapshot)
     }
 
-    const labels: {[label in Labels]?: string} = {}
-    for (const key of keys) {
-      if (isSecondary(key.index)) {
-        labels[key.index.options.label] = key.query()
+    await data.set(primary, snakecase(prune(this), { deep: true }), this.secondaries())
+
+    if (this.shadowKeys) {
+      if (
+        this.shadowKeys().length > MAX_SHADOW_KEYS &&
+        !this.unsafeShadowKeysUnbounded?.()
+      ){
+        throw new Error(`Maximum number of ${MAX_SHADOW_KEYS} shadowKeys exceeded: Reduce number of shadowKeys, or add unsafeShadowKeysUnbounded method to model`)
       }
-    }
 
-    await data.set(primary, snakecase(prune(this), { deep: true }), labels)
+      const keySets = this.shadowKeys()
+      const shadowPrimaries = keySets.map((keySet) =>
+        this.primary(keySet)
+      )
+
+      await Promise.all([
+        ...this.__shadowSnapshots.map((shadowSnapshot) => {
+          if (!shadowPrimaries.includes(shadowSnapshot)) {
+            return data.remove(shadowSnapshot)
+          }
+        }),
+        ...shadowPrimaries.map((shadowPrimary, index) =>
+          data.set(
+            shadowPrimary,
+            snakecase(prune(this), { deep: true }),
+            this.secondaries(keySets[index])
+          )
+        ),
+      ])
+    }
   }
 
   /**
    *
-   * Will delete the model from the database.
+   * Will delete the model and any shadows from the database.
    *
    */
   async delete() {
```

Comment:
This method also has a similar bug to `save()`. When deleting a model, the snapshots (both the main and shadow snapshots) are not updated (should be set to `undefined` I suppose?). To re-iterate, this is not a newly introduced bug, but one that exists in current version as well, but it also affects shadow primaries.